### ISSUE TITLE
nix 2.16 requires meta.position to be an absolute path

### ISSUE THREAD

**Issue by guibou** [NixOS/nix#8662]
I'm using `meta.position` in order to allow the users of our build system to edit the definition of a rule using `nix edit .#ruleName`.

It was working fine in nix `2.15`, but since nix `2.16` it complains about requiring an absolute path:

```
error:
       … while evaluating the 'meta.position' attribute of a derivation

         at «none»:0: (source not available)

       error: string 'foo/bar/baz.nix:54' doesn't represent an absolute path
```

I'm able to provide an absolute path in the value of `meta.position` using an hardcoded absolute path, but I don't know how to generate an absolute path which does not point to the nix store, e.g. `toString ./.` points to the nix store.

This is a problem for us because we are extensively relying on nix to provide a build system for users which are not well used to nix and most of the time they just need to do some simple fixs, and we leveraged `nix edit` in order to open their editor on the right location. This is not possible anymore.

**Steps To Reproduce**

Create a `flake.nix` in directory `/foo/bar/baz` with:

```nix
{
  description = "A very basic flake";

  outputs = { self, nixpkgs }: {

    packages.x86_64-linux.hello = nixpkgs.legacyPackages.x86_64-linux.hello;

    packages.x86_64-linux.default = {
      meta.position = "flake.nix:9";
    };

  };
}
```

```shell
$ nix run --impure nixpkgs#nixUnstable -- --version     
nix (Nix) 2.16.1
$ nix run --impure nixpkgs#nixUnstable -- edit .#default
error:
       … while evaluating the 'meta.position' attribute of a derivation

         at «none»:0: (source not available)

       error: string 'flake.nix:9' doesn't represent an absolute path
       
$ nix run --impure nixpkgs#nix -- --version             
nix (Nix) 2.15.1
$ nix run --impure nixpkgs#nix -- edit .#default 
# Works fine
```

Changing the `meta.position` to `/foo/bar/baz/flake.nix` works fine in both context.

**Comment by Artturin** [comment]
Caused by
https://github.com/NixOS/nix/commit/94812cca98fbb157e5f64a15a85a2b852d289feb#diff-e23eedbf9fd66095408ef60947098af06bbb24dc302c88d895349f71513eb9fcR122
@edolstra

**Comment by roberth** [comment]
I don't think strings can work in the general case, because strings don't remember which flake they came from.
Does this work for your use case?

```
mkDerivation {
  pos = __curPos;
}
```

If you have abstractions on top of mkDerivation, you'll want to pass `pos` along unmodified from the "end user" definition.

**Comment by thufschmitt** [comment]
Discussed during the Nix maintainers meeting on 2024-02-16.
This won't be fixed as it's only affecting the surface CLI, and allowing relative paths won't really make sense with the upcoming source path abstraction. Besides, @roberth's workaround should work.

<details>

- Only happens in `nix edit` (nothing changed in the evaluator itself)
- Working around that would be very hard, and it's very surfacic.
- String does not contain enough information. It really should be a path.
- Wontfix

**Comment by guibou** [comment]
Sorry, I'm unsure I understand the hack with `__curPos`.

Should I move the `pos = __curPos` on ALL my end user definitions?

This is not convenient because it exposes internal to the end user.

I'll wait for the "source path abstraction" before upgrading our users to nix > 2.15.

Thank you for your time.

**Comment by roberth** [comment]
> Should I move the `pos = __curPos` on ALL my end user definitions?

That would certainly fix the problem, but ideally you could pull this info from an attribute that your end users pass in.
For example `mkDerivation` itself [makes a guess](https://github.com/NixOS/nixpkgs/blob/2104e113a01a5e0c0b9313f7956b311919fc6d43/pkgs/stdenv/generic/make-derivation.nix#L174) when `pos` isn't explicitly provided.

> the hack with `__curPos`.

I regard it as a proper language feature, but indeed it is [undocumented](https://github.com/NixOS/nix/issues/10036).

> "source path abstraction" before upgrading our users to nix > 2.15.

2.15 is does not receive fixes anymore; we currently backport down to 2.18. I don't expect #6530 to land before the upcoming one or two NixOS releases either.

> We typically backport only as far back as the Nix version used in the latest NixOS release, which is announced in the [NixOS release notes](https://nixos.org/manual/nixos/stable/release-notes#ch-release-notes).

&mdash; https://nixos.org/manual/nix/stable/release-notes/index.html?highlight=backport#nix-release-notes

**Comment by guibou** [comment]
@roberth I do apologize, but I don't really understand how `__curPos` can help in my context.

I would like to use `nix edit .#path.to.something` in order for my user to edit a rule which just failed for some reason.

I was using `meta.position` with a relative path to the root, using the following logic:

```
     pathRelativeToRoot = x: removePrefix (toString root + "/") (toString x);

      position = let p = builtins.unsafeGetAttrPos "name" args; in
        "${pathRelativeToRoot p.file}:${toString p.line}";
```

Where `root` was the path of my flake (in nix store, but removing the prefix would give me a relative path to hopefully the top of my flake, which is local, and usually user would call `nix` at the toplevel of the path.

With `__curPos`, I still need to set `meta.position` to an absolute path in the nix store, which does not work for me. I'm unsure how I can fix `nix edit` with `__curPos`.

This being said, I just realized that I can write an external program which uses `meta.position` and get the behavior I was used to have. I just need to reeducate my users.
